### PR TITLE
Bump `botocore` dependency specification

### LINKED ISSUE #1229: Bump `botocore` dependency
Similar to https://github.com/aio-libs/aiobotocore/commit/89af505abae609314ea6d2ccced2244636ddb191, as the current constraint is blocking updates of `langchain-aws`, which now depends on boto3 `>=1.35.74`:

https://github.com/langchain-ai/langchain-aws/compare/v0.2.7...v0.2.9

### PR DESCRIPTION
**Pull request by jakob-keller**
### Description of Change
This PR intends to improve general compatibility of `aiobotocore` within the Python ecosystem by bumping the dependency specification of `botocore`, as well as `boto3` and `awscli`.

### Assumptions
[Upstream diff](https://github.com/boto/botocore/compare/1.35.36..1.35.81) contains changes that require adjustments to the aiobotocore codebase.

### Checklist for All Submissions
* [x] I have added change info to [CHANGES.rst](https://github.com/aio-libs/aiobotocore/blob/master/CHANGES.rst)
* [x] If this is resolving an issue (needed so future developers can determine if change is still necessary and under what conditions) (can be provided via link to issue with these details): closes #1229
  * [ ] Detailed description of issue
  * [ ] Alternative methods considered (if any)
  * [ ] How issue is being resolved
  * [ ] How issue can be reproduced
* [ ] If this is providing a new feature  (can be provided via link to issue with these details):
  * [ ] Detailed description of new feature
  * [ ] Why needed
  * [ ] Alternatives methods considered (if any)

### Checklist when updating botocore and/or aiohttp versions

* [x] I have read and followed [CONTRIBUTING.rst](https://github.com/aio-libs/aiobotocore/blob/master/CONTRIBUTING.rst#how-to-upgrade-botocore)
* [x] I have updated test_patches.py where/if appropriate (also check if no changes necessary)
* [x] I have ensured that the awscli/boto3 versions match the updated botocore version

### CODE REVIEW

**Comment by jakob-keller** [comment]
Notes to reviewers:

* I believe that [upstream](https://github.com/boto/botocore/compare/1.35.36..1.35.49) does not affect any codepaths relevant for us, beyond what is included in this PR. However, I am not 100% sure about `_handle_200_error()` in `handlers.py`. Please take a closer look at that aspect, if you can.
* FWIW: test coverage is not actually any worse than it was before.

**Comment by thehesiod** [comment]
> Notes to reviewers:
> 
> * I believe that [upstream](https://github.com/boto/botocore/compare/1.35.36..1.35.49) does not affect any codepaths relevant for us, beyond what is included in this PR. However, I am not 100% sure about `_handle_200_error()` in `handlers.py`. Please take a closer look at that aspect, if you can.
> * FWIW: test coverage is not actually any worse than it was before.

ya I'm not sure if `response_dict['body']` is a coroutine or the actual bytes.  I'd put a breakpoint and see what it looks like

**Comment by thehesiod** [comment]
I think it's populated here: https://github.com/aio-libs/aiobotocore/blob/8a50b61c45674a0c1edf9507caf094723c9aad4b/aiobotocore/endpoint.py#L57-L65 so I think it depends on the response type

**Comment by jakob-keller** [comment]
> I think it's populated here:
> 
> https://github.com/aio-libs/aiobotocore/blob/8a50b61c45674a0c1edf9507caf094723c9aad4b/aiobotocore/endpoint.py#L57-L65
> 
> so I think it depends on the response type

Exactly: Case 1 and 4 result in bytes, case 3 cannot be reached in this context, but case 2 on line 60 is tricky. I have no idea under what conditions if at all this is reached.

**Comment by thehesiod** [comment]
I think here's an example: https://botocore.amazonaws.com/v1/documentation/api/latest/reference/eventstream.html

**Comment by thehesiod** [comment]
isn't line 63 also a coroutine? That's like a s3 get-object body i think

**Comment by thehesiod** [comment]
ideally we could add tests for this with moto

**Comment by thehesiod** [comment]
btw you should add yourself to the sponsors, you've already done a ton of work

**Comment by jakob-keller** [comment]
> isn't line 63 also a coroutine? That's like a s3 get-object body i think

There's a guard function `_should_handle_200_error()` that makes that codepath impossible to reach, as far as I can tell.

**Comment by jakob-keller** [comment]
> btw you should add yourself to the sponsors, you've already done a ton of work

Thanks, but I'm good :-)

**Comment by jakob-keller** [comment]
> I think here's an example: https://botocore.amazonaws.com/v1/documentation/api/latest/reference/eventstream.html

> ideally we could add tests for this with moto

I could give it a shot towards the end of the week. Thanks for looking into it!

**Comment by jakob-keller** [comment]
> > isn't line 63 also a coroutine? That's like a s3 get-object body i think
> 
> There's a guard function `_should_handle_200_error()` that makes that codepath impossible to reach, as far as I can tell.

Actually, I am suspecting that this guard function also prevents case 2 / line 60 from being reached.

**Comment by jakob-keller** [comment]
Sorry for the delay. I am hoping to perform the pending tests in the coming days.

**Comment by yoavk** [comment]
> Sorry for the delay. I am hoping to perform the pending tests in the coming days.

@jakob-keller Any chance you can update the dependency to the newest botocore (1.35.76) before running the tests?

**Comment by yoavk** [comment]
> > > Sorry for the delay. I am hoping to perform the pending tests in the coming days.
> > 
> > 
> > @jakob-keller Any chance you can update the dependency to the newest botocore (1.35.76) before running the tests?
> 
> No, I'm afraid not. Also, this PR will not bump the dependency to the current release of botocore. That will require further analysis and likely work on our side, given the amount of product releases and news announced at AWS re:Invent.

Oh well, I was hoping it would be an easy upgrade. I'll double check to see if this version already includes what we need. 
Thanks!

**Comment by jakob-keller** [comment]
I've included all pending botocore updates into this PR, including support for DSQL auth and all the other new and updated AWS services.

We still need include unit tests, in particular with regards to the issue discussed above. Hopefully this weekend...

**Comment by jakob-keller** [comment]
> I think here's an example: https://botocore.amazonaws.com/v1/documentation/api/latest/reference/eventstream.html

Well, that example doesn't work for me: [AWS no longer accepts new users for S3 Select](https://aws.amazon.com/blogs/storage/how-to-optimize-querying-your-data-in-amazon-s3/). I will need to find another service that uses event streams.

**Comment by thehesiod** [comment]
looks like bedrock supports eventstreams as well: https://docs.aws.amazon.com/bedrock/latest/userguide/agents-test.html

**Comment by thehesiod** [comment]
stream example: https://stackoverflow.com/questions/78378549/botocore-eventstream-invalidheaderslength

**Comment by thehesiod** [comment]
oh duh, we should just port their tests, lemme c if I can do it

**Comment by thehesiod** [comment]
@jakob-keller see latest change to https://github.com/aio-libs/aiobotocore/pull/1232.  Wasn't sure how to push to your branch.  If you give me rights can push there or feel free to cherry pick it

**Comment by thehesiod** [comment]
seems like it's fairly easy to pull their tests in now, perhaps we should try to pull the whole thing over

**Comment by jakob-keller** [comment]
> seems like it's fairly easy to pull their tests in now, perhaps we should try to pull the whole thing over

That sounds fantastic, thanks!

I have to admit I'm still not familiar with our test suite. Wouldn't it make sense to separate your test related work into its own PR and then revisit this PR?

**Comment by jakob-keller** [comment]
#1234 cherry picks test improvements by @thehesiod and then some. Once that is merged, I will do a rebase and finalize this PR.

**Comment by thehesiod** [comment]
oo you got the magic number :)

**Comment by thehesiod** [comment]
sure, I'm not too picky about "clean" prs

**Comment by jakob-keller** [comment]
> seems like it's fairly easy to pull their tests in now, perhaps we should try to pull the whole thing over

I've included relevant unit tests from upstream, resulting in full test coverage.

**Comment by jakob-keller** [comment]
> seems like it's fairly easy to pull their tests in now, perhaps we should try to pull the whole thing over

You mean, the entire botocore test suite? That should be a separate issue for sure ;-)

**Comment by thehesiod** [comment]
> > seems like it's fairly easy to pull their tests in now, perhaps we should try to pull the whole thing over
> 
> You mean, the entire botocore test suite? That should be a separate issue for sure ;-)

yes of course =)

**Review by thehesiod — APPROVED** [review]

**Comment by thehesiod** [comment]
sorry I thought I approved this yesterday

**Comment by thehesiod** [comment]
oh, shoot forgot to see if we got coverage now for that new method

**Comment by jakob-keller** [comment]
> oh, shoot forgot to see if we got coverage now for that new method

Patch coverage is not relevant here, since the method remains in botocore and is not replaced by us.

I was not able to come up with an example that triggers that method locally. I tried the S3 Select API with and without moto and for some reason the breakpoints never hit. Probably my bad.

Having analyzed the code path, I would be surprised if we have an issue, but still can't fully rule it out. Looking at possible values for `response_dict['body']` there are four cases, as you rightly pointed out:


> I think it's populated here:
> 
> https://github.com/aio-libs/aiobotocore/blob/8a50b61c45674a0c1edf9507caf094723c9aad4b/aiobotocore/endpoint.py#L57-L65
> 
> so I think it depends on the response type

* case 1 results in bytes
* case 2 cannot be reached in this case
* case 3 might be reached, but I don't know how that could be handled. We are checking for an error message encoded in the response body. We would need to consume the streaming response to perform that check. I assume it would then be unavailable for consumption by the user?
* case 4 results in bytes

**Review by thehesiod — APPROVED** [review]

**Comment by thehesiod** [comment]
welp lets get it out, if something breaks we can figure out what tests we need to port over or new ones to write

File: requirements-dev.in [hunk 1]
```diff
@@ -22,4 +22,5 @@ docker~=7.1
 moto[server,s3,sqs,awslambda,dynamodb,cloudformation,sns,batch,ec2,rds]~=4.2.9
 pre-commit~=3.5.0
 pytest-asyncio~=0.23.8
+time-machine~=2.15.0
```
Comment (thehesiod) [inline requirements-dev.in:25]: this doesn't appear to have been needed

**Pull request merged**